specs/core/0_fork-choice.md

```diff
@@ -34,76 +38,131 @@ All terminology, constants, functions, and protocol mechanics defined in the [Ph
-* Let `get_children(store: Store, block: BeaconBlock) -> List[BeaconBlock]` return the child blocks of the given `block`.
-* Let `justified_head_state` be the resulting `BeaconState` object from processing the chain up to the `justified_head`.
-* The `head` is `lmd_ghost(store, justified_head_state, justified_head)` where the function `lmd_ghost` is defined below. Note that the implementation below is suboptimal; there are implementations that compute the head in time logarithmic in slot count.
+#### `get_head`
 
 ```python
-def lmd_ghost(store: Store, start_state: BeaconState, start_block: BeaconBlock) -> BeaconBlock:
-    """
-    Execute the LMD-GHOST algorithm to find the head ``BeaconBlock``.
-    """
-    validators = start_state.validator_registry
-    active_validator_indices = get_active_validator_indices(validators, slot_to_epoch(start_state.slot))
-    attestation_targets = [(i, get_latest_attestation_target(store, i)) for i in active_validator_indices]
-
-    # Use the rounded-balance-with-hysteresis supplied by the protocol for fork
-    # choice voting. This reduces the number of recomputations that need to be
-    # made for optimized implementations that precompute and save data
-    def get_vote_count(block: BeaconBlock) -> int:
-        return sum(
-            start_state.validator_registry[validator_index].effective_balance
-            for validator_index, target in attestation_targets
-            if get_ancestor(store, target, block.slot) == block
-        )
-
-    head = start_block
-    while 1:
-        children = get_children(store, head)
+def get_head(store: Store) -> Bytes32:
+    # Execute the LMD-GHOST fork choice
+    head = store.justified_root
+    while True:
+        children = [root for root in store.blocks.keys() if store.blocks[root].parent_root == head]
         if len(children) == 0:
             return head
-        # Ties broken by favoring block with lexicographically higher root
-        head = max(children, key=lambda x: (get_vote_count(x), hash_tree_root(x)))
+        # Sort by attesting balance with ties broken lexicographically
+        head = max(children, key=lambda root: (get_attesting_balance_from_store(store, root), root))
+```
+
+### Handlers
+
+#### `on_tick`
+
+```python
+def on_tick(store: Store, time: int) -> None:
+    store.time = time
 ```
 
-## Implementation notes
+#### `on_block`
 
-### Justification and finality at genesis
+```python
+def on_block(store: Store, block: BeaconBlock) -> None:
+    # Check block is a descendant of the finalized block
+    assert get_ancestor(store, signing_root(block), store.blocks[store.finalized_root].slot) == store.finalized_root
+    # Check block slot against Unix time
+    pre_state = store.states[block.parent_root]
+    assert store.time >= pre_state.genesis_time + block.slot * SECONDS_PER_SLOT
+    # Check the block is valid and compute the post-state
+    state = state_transition(pre_state, block)
```

Comment:
Ahhhh a mutable issue here! 💣
1. `state_transition` returns a `state`, but note that there is side effect - the given `state` input would be changed anyway. (#1059)
2. Since `store.states` is a `dict`, the updated `pre_state` would mess up `store.states`.

A quick fix for this PR might still be adding the scary `deepcopy` inside `state_transition` that I know [people really don't want to have in the spec](https://github.com/ethereum/eth2.0-specs/issues/1059#issuecomment-490162991). :'(

But open to continue discussing in #1059.